qiskit_experiments/framework/json.py

```diff
@@ -0,0 +1,516 @@
+        its class as ``cls()``.
+
+    .. note::
+
+        Serialization of custom classes works for user-defined classes in
+        Python scripts, notebooks, or third party modules. Note however
+        that these will only be able to be de-serialized if that class
+        can be imported form the same scope at the time the
+        :class:`ExperimentDecoder` is invoked.
+    """
+
+    def default(self, obj: Any) -> Any:  # pylint: disable=arguments-differ
+        if isinstance(obj, complex):
+            return _serialize_safe_float(obj)
+        if isinstance(obj, set):
+            return {"__type__": "set", "__value__": list(obj)}
+        if isinstance(obj, np.ndarray):
+            value = _serialize_and_encode(obj, np.save, allow_pickle=False)
+            return {"__type__": "ndarray", "__value__": value}
+        if isinstance(obj, sps.spmatrix):
+            value = _serialize_and_encode(obj, sps.save_npz, compress=False)
+            return {"__type__": "spmatrix", "__value__": value}
+        if isinstance(obj, bytes):
+            return _serialize_bytes(obj)
+        if dataclasses.is_dataclass(obj):
+            return _serialize_object(obj, settings=dataclasses.asdict(obj))
+        if isinstance(obj, QuantumCircuit):
+            # TODO Remove the decompose when terra 6713 is released.
+            if isinstance(obj, BlueprintCircuit):
+                obj = obj.decompose()
+            value = _serialize_and_encode(
+                data=obj, serializer=lambda buff, data: qpy_serialization.dump(data, buff)
+            )
+            return {"__type__": "QuantumCircuit", "__value__": value}
+        if isinstance(obj, ParameterExpression):
+            value = _serialize_and_encode(
+                data=obj,
+                serializer=qpy_serialization._write_parameter_expression,
+                compress=False,
+            )
+            return {"__type__": "ParameterExpression", "__value__": value}
+        if isinstance(obj, Result):
+            return {"__type__": "Result", "__value__": obj.to_dict()}
+        if isinstance(obj, QuantumChannel):
+            # Temporary fix for incorrect settings in qiskit-terra
+            # See https://github.com/Qiskit/qiskit-terra/pull/7194
+            settings = {
+                "data": obj.data,
+                "input_dims": obj.input_dims(),
+                "output_dims": obj.output_dims(),
+            }
+            return _serialize_object(obj, settings=settings)
+        if isinstance(obj, DensityMatrix):
+            # Temporary fix for incorrect settings in qiskit-terra
+            # See https://github.com/Qiskit/qiskit-terra/pull/7194
+            settings = {
+                "data": obj.data,
+                "dims": obj.dims(),
+            }
+            return _serialize_object(obj, settings=settings)
+        if is_type(obj):
+            return _serialize_type(obj)
+        try:
+            return super().default(obj)
```

Comment:
What happens if we serialize a python dict that contains a byte object? Is the method recursively called for each value?

Reply:
Yes, thats how JSON encoding works.